### ISSUE TITLE
Animation Sequencer

### ISSUE THREAD

**Issue by astillich** [REGoth-project/REGoth#82]
Create a class which coordinates high-level animation playback:

* handle animation layers: replace animation / queue animation so that it is played when the current animation on the layer is finished
* blend between animations
* switch models
* spawn particle FX
* play sounds

All 'foreign' functionality (like particles / sound) should be delegated to the respective dedicated systems. Use messages to create a loose coupling between them (imperative style: snd_play, pfx_spawn, etc).

Structure could look like this .. each entity has a sequencer, probably wrapped in an AnimationController component which handles per-entity state:

* AnimationSequencer
** AnimationLayer
*** SkeletalAnimationTrack
*** ParticleAnimationTrack
*** SoundAnimationTrack
*** TagAnimationTrack
** AnimationLayer
** ...

Use the current AnimHandler class as a basis to handle skeletal animation (it could be refactored to SkeletalAnimationTrack).

Please leave comments, additions, suggestions.

**Comment by astillich** [comment]
@degenerated1123 I think you mentioned that you had some blending code? Does that also handle layering? Would be nice to have a look before starting this.

**Comment by astillich** [comment]
AnimationSequencer instances (the setup read from model scripts) could be shared between entities. The AnimationController would then have a per-entity AnimationState instance which keeps track of the sequencer state for that entity.

**Comment by degenerated1123** [comment]
There is currently no blending whatsoever, nor layering. However, I'm not sure whether gothic even used "layering", as in automatically done. They did have their "animation-overlays", which replace some of the animations of the base animation-set. But these overlays always contain an animation for the full skeleton, if I'm not mistaken.

**Comment by astillich** [comment]
Just had a quick look at the G2 .MDS files: it at last uses layer 1 and 20.

**Comment by degenerated1123** [comment]
You're probably right, I don't think they duplicated every single animation just because of maybe holding torch. Makes more sense to just overwrite the nodes used by the torch-animation.
Will implement this into the AnimationHandler!

**Comment by Czarek22089** [comment]
Quick question, does moving characters in REgoth will be using root motion in animation to move them like in original ? Becouse from what I saw when testing, moving was just from simple script... hope you understand

**Comment by degenerated1123** [comment]
That is working, yes. At least for the player it is, NPCs use something simpler for now.

**Comment by degenerated1123** [comment]
This is implemented now.
